### PR TITLE
Elastic this, Elastic that

### PR DESCRIPTION
**Pull request by jstange**
AWS Elasticsearch
-------------------

Full API coverage for Amazon's Elasticsearch-as-a-service offering. The resource type is called `SearchDomain`.

http://54.175.86.194/docs/MU/Config/BasketofKittens/search_domains.html

The following is a legal BoK:

```
---
  appname: searchdomain
  search_domains:
  - name: test
    instance_type: t2.small.elasticsearch
```

A declaration that exercises (most of) the esoteric functionality:

```
  search_domains:
  - name: testcomplex
    instance_type: t2.small.elasticsearch
    instance_count: 4
    dedicated_masters: 1
    master_instance_type: t2.medium.elasticsearch
    zone_aware: true
    ebs_iops: 100
    ebs_type: gp2
    elasticsearch_version: "6.2"
    advanced_options:
      "indices.query.bool.max_clause_count": 512
    cognito:
      identity_pool_id: "us-east-1:69e2223c-2c74-42ca-9b27-1037fcb60b87"
      user_pool_id: "us-east-1_eSwWA1VYQ"
    slow_logs: MY-INDEXERS-ARE-SLOW-TELL-CLOUDWATCH
    vpc:
      vpc_name: vpc
```

Note that you can deploy an Elasticsearch domain *without* specifying a VPC. This will, however, put its endpoints on the public internet and security groups won't be an available feature. So you probably want to stay in VPC world.

Other cloud providers do not currently have a managed Elasticsearch service, so this is unlikely to have other implementations in other layers. Most of its configuration language is very particular to AWS.

That aside, this implementation is clean and fresh example of how to implement a cloud resource "right." Of particular note:

- Cloud-specific validation (which is 90% of this particular thing) happens in `MU::Cloud::AWS::SearchDomain` instead of somewhere under `MU::Config`
- Validation checks for the existence of other AWS resources (Cognito, CloudWatch) if you specify them.
- A `groom` on an existing `SearchDomain`, e.g. when invoked by `mu-deploy -u DEPLOY-ID -l`, will update any/all modifiable configuration parameters of the existing resource.
- `#notify` is fully implemented, which we've been bad about in the past.

Developers can look at these files in particular as a guiding star:

```
modules/mu/config/search_domain.rb
modules/mu/cloud.rb
modules/mu/clouds/aws.rb
modules/mu/clouds/aws/search_domain.rb
```

AWS Simple Queue Service
---------------------------

Did SQS, too. I had to name the resource `MsgQueue` internally, because the `Queue` namespace is used by Ruby's threading, and while it's not technically a reserved word, that ends up being way too annoying to work around.

http://54.175.86.194/docs/MU/Config/BasketofKittens/msg_queues.html

Here's a BoK:

```
---
appname: queues
msg_queues:
- name: a_queue
```

Here's a more complicated BoK:

```
---
appname: queues
msg_queues:
- name: thing1
  visibility_timeout: "1 hour"
  failqueue:
    create: true
    retries_before_fail: 5
- name: thing2
  retain: 2d
  delay: "30 seconds"
#  kms:
#    key_id: bf64a093-2c3d-46fa-0d4f-8232fa7ed53 # not a real key
- name: thing3
  fifo: true
  max_msg_size: 2
- name: thing4
  dedup: true
```

There is one field that's not implemented, for defining an IAM policy for access to a queue. The documentation is incomplete and it's not obvious what valid input is, so I'm letting it be for now. We need to revisit how we interact with IAM anyway. See https://jira.egt-labs.com/browse/CLOUDMATIC-135

AWS Elastic Container Service
------------------------------

Sort of. The project that would have made use of this doesn't appear to benefit from containers, so I've stopped work after getting as far as building a container cluster with instances registered. The resource type is called `ContainerCluster`.

http://54.175.86.194/docs/MU/Config/BasketofKittens/container_pools.html

Technically you could use that to build an empty ECS cluster, and then plant things there on your own time. Eh.

ECS, EKS (Elastic Kubernetes Service), and Google's GKE all appear to be structured in a similar fashion. An Autoscale group is spun up and an agent installed on each node, and containers are deployed as a secondary behavior. AWS Fargate is slightly different, in that instance creation is handled by Amazon.

A completed implementation of `ContainerCluster` would support EKS, Fargate, GKE, and provide a mechanism for container deployment both at create time and later. We would perhaps even bundle a Docker Registry service with Mu.

More MU::Config refactoring
----------------------------

Following the migration of cloud-specific configuration behaviors into the relevant `MU::Cloud` implementations, I've further broken up the monolithic general schema and validation routines from `MU::Config` into resource class-specific chunks. See `modules/mu/config/*.rb`. This makes particular pieces of config processing much easier to find, and the addition of new resource types a cleaner affair. The README files for cloud layer developers have been updated accordingly.

This particular change merits broad testing. If `mu-deploy -d` works on an existing BoK with the current `master` but not on this branch, there might be a bug to chase.

Bugfixes
--------------------------------------

- A possible race condition & crash in `MU::MommaCat#syncLitter` should be fixed
- IAM role deletion should now work properly for roles which have attached canned AWS policies
- IAM role deletion should now work properly for roles in accounts with large numbers of IAM roles
- `MU::Cloud::AWS::Log.find` and `MU::Cloud::AWS::Log#create` are now more robust
- Cost-calculation pseudo-deploys should stop leaving detritus in `/opt/mu/var/deployments`
- References to defunct S3 buckets for our Ruby RPMs and some Powershell support libraries have been excised
- Dangling `FirewallRule` cleanup issues should be fixed
- Harmless-but-annoying threading weirdness around cost calculator deploys maybe has a workaround

### CODE REVIEW

**Comment by jstange** [comment]
Took too long, now SQS is in here too.

File: modules/mu/mommacat.rb [hunk 1]
```diff
@@ -2075,35 +2088,38 @@ def syncLitter(nodeclasses = [], triggering_node: nil, save_all_only: false)
...
-              end
-            end
-          }
+        litter = {}
+        nodeclasses.each { |nodeclass|
+          litter.merge!(findLitterMate(type: "server", name: nodeclass, return_all: true))
+        }
+        litter.each_pair { |mu_name, node|
+          next if !triggering_node.nil? and mu_name == triggering_node.mu_name
+          if !node.deploydata.keys.include?('nodename')
```
Comment (ryantiger658) [inline modules/mu/mommacat.rb]: @jstange I am getting a bug here. When I attempt to deploy a server pool everything deploys and grooms properly then I get this error, and it purges my deployment.

```May 14 15:43:24 - groomers/chef - Adding recipe[ipaws-capaggregator] to Chef run_list of IPAWS-DEV-2018051415-BY-CAPAGGREGATOR-YA0
May 14 15:43:25 - groomers/chef - Saving IPAWS-DEV-2018051415-BY-CAPAGGREGATOR-YA0 Chef artifacts
May 14 15:43:26 - deploy - #<NoMethodError: undefined method `keys' for nil:NilClass>
["/opt/mu/lib/modules/mu/mommacat.rb:2118:in `block in syncLitter'",
 "/opt/mu/lib/modules/mu/mommacat.rb:2116:in `each_pair'",
 "/opt/mu/lib/modules/mu/mommacat.rb:2116:in `syncLitter'",
 "/opt/mu/lib/modules/mu/mommacat.rb:2478:in `save!'",
 "/opt/mu/lib/modules/mu/mommacat.rb:1326:in `notify'",
 "/opt/mu/lib/modules/mu/cloud.rb:1344:in `block (4 levels) in <class:Cloud>'",
 "/opt/mu/lib/modules/mu/clouds/aws/server_pool.rb:440:in `block (2 levels) in create'",
 "/opt/rubies/ruby-2.3.1/lib/ruby/gems/2.3.0/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'"]```

**Comment by ryantiger658** [comment]
When I try to run a `msg_queue` I get the following error:

```/opt/rubies/ruby-2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- chronic_duration (LoadError)
        from /opt/rubies/ruby-2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from /opt/mu/lib/modules/mu/clouds/aws/msg_queue.rb:332:in `validateConfig'
        from /opt/mu/lib/modules/mu/config.rb:998:in `insertKitten'
        from /opt/mu/lib/modules/mu/config.rb:1489:in `block (2 levels) in validate'
        from /opt/mu/lib/modules/mu/config.rb:1488:in `each'
        from /opt/mu/lib/modules/mu/config.rb:1488:in `block in validate'
        from /opt/mu/lib/modules/mu/config.rb:1487:in `each'
        from /opt/mu/lib/modules/mu/config.rb:1487:in `validate'
        from /opt/mu/lib/modules/mu/config.rb:612:in `initialize'
        from /opt/mu/bin/mu-deploy:91:in `new'
        from /opt/mu/bin/mu-deploy:91:in `<main>'```

**Comment by ryantiger658** [comment]
When I try to run the complex search_domain example I get the following error:

```May 14 21:36:45 - clouds/aws/search_domain - Specified slow_logs CloudWatch log group 'superBoK_logs' in SearchDomain 'search_domain-complex' doesn't appear to exist
/opt/mu/lib/modules/mu/clouds/aws/search_domain.rb:321:in `validateConfig': undefined method `arn' for nil:NilClass (NoMethodError)
        from /opt/mu/lib/modules/mu/config.rb:998:in `insertKitten'
        from /opt/mu/lib/modules/mu/config.rb:1489:in `block (2 levels) in validate'
        from /opt/mu/lib/modules/mu/config.rb:1488:in `each'
        from /opt/mu/lib/modules/mu/config.rb:1488:in `block in validate'
        from /opt/mu/lib/modules/mu/config.rb:1487:in `each'
        from /opt/mu/lib/modules/mu/config.rb:1487:in `validate'
        from /opt/mu/lib/modules/mu/config.rb:612:in `initialize'
        from /opt/mu/bin/mu-deploy:91:in `new'
        from /opt/mu/bin/mu-deploy:91:in `<main>'```

Even though `superBoK_logs` is a defined resource in my BoK.

File: modules/mu/mommacat.rb (hunk 1 above)
Comment (jstange) [reply modules/mu/mommacat.rb]: Looks trivial enough, and actually not from this branch's code. Adding a fix.

**Comment by jstange** [comment]
@ryantiger658 Your `chronic_duration` error looks like the `Gemfile` wasn't honored or bundle wasn't updated or something. Did you switch to this branch with `mu-self-update`, or just a checkout?

**Comment by ryantiger658** [comment]
I likely just did a checkout. I assume `mu-self-update` will flush out my gem stuffs?

File: modules/mu/mommacat.rb (hunk 1 above)
Comment (ryantiger658) [reply modules/mu/mommacat.rb]: I looked at the changelog, and there were a bunch of changes to this chunk of code included, so I just assumed.

**Comment by jstange** [comment]
@ryantiger658 Yeah, `mu-self-update` will sync up bundles and whatnot.

Usually a checkout is an ok shortcut, but I happened to add a gem to the ecosystem to support some magical config parser behavior this time around.

**Pull request merged**